tests/unit/services/AdvancedSearchQueryTest.php

```diff
@@ -0,0 +1,45 @@
+<?php declare(strict_types=1);
+/**
+ * @author Nicolas CARPi <nico-git@deltablot.email>
+ * @author Marcel Bolten <github@marcelbolten.de>
+ * @copyright 2012 Nicolas CARPi
+ * @see https://www.elabftw.net Official website
+ * @license AGPL-3.0
+ * @package elabftw
+ */
+
+namespace Elabftw\Services;
+
+class AdvancedSearchQueryTest extends \PHPUnit\Framework\TestCase
+{
+    public function testGetWhereClause(): void
+    {
+        $query = ' TEST TEST1 AND TEST2 OR TEST3 NOT TEST4 & TEST5';
+        $query .= ' | TEST6 AND ! TEST7 (TEST8 or TEST9) "T E S T 1 0"';
+        $query .= ' \'T E S T 1 1\' "chinese 汉语 漢語 中文" "japanese 日本語 ひらがな 平仮名 カタカナ 片仮名" ';
+
+        $advancedSearchQuery = (new AdvancedSearchQuery($query, 'body'));
+        $whereClause = $advancedSearchQuery->getWhereClause();
+        $this->assertIsArray($whereClause);
+        $this->assertStringStartsWith(' AND ((entity.body LIKE :', $whereClause['where']);
+        $this->assertStringEndsWith(')))', $whereClause['where']);
+    }
+
+    public function testSyntaxError(): void
+    {
+        $query = 'AND AND AND';
+
+        $advancedSearchQuery = (new AdvancedSearchQuery($query, 'body'));
+        $advancedSearchQuery->getWhereClause();
+        $this->assertStringStartsWith('Column ', $advancedSearchQuery->getException());
+    }
+
+    public function testComplexityLimit(): void
+    {
+        $query = 'TEST TEST1';
+
+        $advancedSearchQuery = (new AdvancedSearchQuery($query, 'body', 1));
+        $advancedSearchQuery->getWhereClause();
+        $this->assertEquals('Query is to complex.', $advancedSearchQuery->getException());
```

Comment:
how is "TEST TEST1" too complex?

Reply:
btw missing an "o" there ;)

Reply:
It is too complex in this test because of the last parameter "1" of AdvancedSearchQuery. It defines a depth limit.
The idea is to be able to define a maximum complexity to avoid very long/"nested" where clauses which could reduce the performance. It is actually not used so far for the user input so far.
I will fix the typo later today.